pkg/services/ngalert/state/manager.go

```diff
@@ -164,18 +163,50 @@ func (st *Manager) Get(orgID int64, alertRuleUID, stateId string) *State {
 	return st.cache.get(orgID, alertRuleUID, stateId)
 }
 
-// ResetStateByRuleUID deletes all entries in the state manager that match the given rule UID.
-func (st *Manager) ResetStateByRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey) []*State {
+// ResetStateByDeletedRuleUID deletes all entries in the state manager that match the given rule UID.
+func (st *Manager) ResetStateByDeletedRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey) []*State {
+	return st.resetStateByRuleUID(ctx, ruleKey, nil, ngModels.StateReasonDeleted)
+}
+
+// ResetStateByPausedRuleUID deletes all entries in the state manager that match the given rule UID.
+func (st *Manager) ResetStateByPausedRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey, rule *ngModels.AlertRule) []*State {
+	return st.resetStateByRuleUID(ctx, ruleKey, rule, ngModels.StateReasonPaused)
+}
+
+func (st *Manager) resetStateByRuleUID(ctx context.Context, ruleKey ngModels.AlertRuleKey, rule *ngModels.AlertRule, reason string) []*State {
```

Comment:
Please add a comment here to explain that `rule` should be nil when reason is `StateReasonDeleted`, and non-nil for `StateReasonPaused`.